smalltalksrc/VMMaker/StackInterpreterPrimitives.class.st

```diff
@@ -953,71 +953,77 @@ StackInterpreterPrimitives >> primitiveDoPrimitiveWithArgs [
 	 or Context>>receiver: anObject tryPrimitive: primIndex withArgs: argArray.
 	 If this primitive fails, arrange that its error code is a negative integer, to
 	 distinguish between this failing and the primitive it invokes failing."
-	| argumentArray arraySize index primIdx savedNumArgs rcvr |
+
+	| argumentArray arraySize index primIdx savedNumArgs rcvr savedTempOop savedTempOop2 |
 	"See checkForAndFollowForwardedPrimitiveState"
 	metaAccessorDepth := -2.
-	(argumentCount between: 2 and: 3) ifFalse:
-		[^self primitiveFailFor: PrimErrUnsupported negated].
+	(argumentCount between: 2 and: 3) ifFalse: [
+		^ self primitiveFailFor: PrimErrUnsupported negated ].
 	argumentArray := self stackTop.
 	primIdx := self stackValue: 1.
-	((objectMemory isArray: argumentArray)
-	 and: [objectMemory isIntegerObject: primIdx]) ifFalse:
-		[^self primitiveFailFor: PrimErrBadArgument negated].
+	((objectMemory isArray: argumentArray) and: [
+		 objectMemory isIntegerObject: primIdx ]) ifFalse: [
+		^ self primitiveFailFor: PrimErrBadArgument negated ].
 	arraySize := objectMemory numSlotsOf: argumentArray.
-	(self roomToPushNArgs: arraySize) ifFalse:
-		[^self primitiveFailFor: PrimErrLimitExceeded negated].
+	(self roomToPushNArgs: arraySize) ifFalse: [
+		^ self primitiveFailFor: PrimErrLimitExceeded negated ].
 
 	primIdx := objectMemory integerValueOf: primIdx.
 	primitiveFunctionPointer := self functionPointerFor: primIdx.
-	primitiveFunctionPointer = 0 ifTrue:
-		[primitiveFunctionPointer := #primitiveDoPrimitiveWithArgs.
-		 ^self primitiveFailFor: PrimErrBadIndex negated].
+	primitiveFunctionPointer = 0 ifTrue: [
+		primitiveFunctionPointer := #primitiveDoPrimitiveWithArgs.
+		^ self primitiveFailFor: PrimErrBadIndex negated ].
 
 	"Pop primIndex and argArray, then push args in place..."
 	(savedNumArgs := argumentCount) = 3
-		ifTrue: "...and receiver if the three arg form"
-			[tempOop2 := self stackValue: 4. "actual receiver"
-			 rcvr := self stackValue: 3. "receiver for primitive"
-			 (objectMemory isOopForwarded: rcvr) ifTrue:
-				[rcvr := objectMemory followForwarded: rcvr].
-			 self pop: 4; push: rcvr] "use first arg as receiver"
-		ifFalse:
-			[self pop: 2].
+		ifTrue: [ "use first arg as receiver""...and receiver if the three arg form"
+			savedTempOop2 := tempOop2.
+			tempOop2 := self stackValue: 3. "actual receiver"
+			rcvr := self stackValue: 2. "receiver for primitive"
+			(objectMemory isOopForwarded: rcvr) ifTrue: [
+				rcvr := objectMemory followForwarded: rcvr ].
```

Comment:
As a nice side effect, fixing the indexes here (0 based) "fixes" the following test in Pharo:

```
MirrorPrimitivesTest >> testExecutingPrimitive
	| actual |
	<expectedFailure> "it will be supported by VM at some point"

	actual := MirrorPrimitives withReceiver: 100 tryPrimitive: 1 withArguments: #(2).

	self assert: actual equals: 102
```